drv/ignition-api/src/lib.rs

```diff
@@ -33,105 +36,395 @@ pub const PORT_MAX: usize = 40;
 )]
 pub enum IgnitionError {
     ServerDied = 1,
+    /// Indicates an error communicating with the FPGA implementing the
+    /// Controller.
     FpgaError,
+    /// Indicates the given port number is larger than the `port_count`.
     InvalidPort,
+    /// Indicates an invalid value was provided.
     InvalidValue,
+    /// Indicates no Target is present/connected to the port.
     NoTargetPresent,
+    /// Indicates the Target is already executing a request. Poll the Target
+    /// state after some time and retry if desired.
     RequestInProgress,
+    /// Indicates the given request conflicts with the Target system power
+    /// state. Poll the Target state and retry if desired.
+    RequestDiscarded,
 }
 
-bitfield! {
-    #[derive(Copy, Clone, Debug, Default, PartialEq, Eq, FromPrimitive, From, FromBytes, AsBytes)]
-    #[repr(C)]
-    pub struct PortState(u64);
-    pub target_present, _: 0;
-    pub u8, into ReceiverStatus, receiver_status, _: 15, 8;
-    u64, into Target, raw_target, _: 63, 16;
+impl From<ServerDeath> for IgnitionError {
+    fn from(_e: ServerDeath) -> Self {
+        Self::ServerDied
+    }
 }
 
-impl PortState {
-    pub fn target(&self) -> Option<Target> {
-        if self.target_present() {
-            Some(self.raw_target())
-        } else {
-            None
+impl From<FpgaError> for IgnitionError {
+    fn from(_e: FpgaError) -> Self {
+        Self::FpgaError
+    }
+}
+
+/// `Ignition` aims to provide a more abstracted and stable API for consumers of
+/// the Ignition subsystem than the implementation specific data provided by
+/// `drv-ignition-server`.
+pub struct Ignition {
+    controller: idl::Ignition,
+}
+
+impl Ignition {
+    pub fn new(task_id: userlib::TaskId) -> Self {
+        Self {
+            controller: idl::Ignition::from(task_id),
+        }
+    }
+
+    /// Return the number of active Controller ports. This value is expected to
+    /// be 35 for production systems but may be smaller in development
+    /// environments. See the note above on `PORT_MAX` about the upper bound.
+    #[inline]
+    pub fn port_count(&self) -> Result<u8, IgnitionError> {
+        self.controller.port_count()
+    }
+
+    /// Return a u64 with each bit indicating whether or not a Target is present
+    /// for this port. See the note above on `PORT_MAX` about the upper bound.
+    #[inline]
+    pub fn presence_summary(&self) -> Result<u64, IgnitionError> {
+        self.controller.presence_summary()
+    }
+
+    /// Return the raw state for the given port. The internals of `PortState`
+    /// are purposefully opague since they are closely tied to the RTL
+    /// implementation and as such are not intended for direct consumption. The
+    /// `port(..)` and `target(..)` functions below are probably what you want
+    /// to use.
+    #[inline]
+    pub fn raw_port_state(&self, port: u8) -> Result<PortState, IgnitionError> {
```

Comment:
Similar to a comment Matt left elsewhere - if the details of `PortState` are opaque, does this method need to exist at all? When would a caller want this over `port()`?

Reply:
Well, `Port` initially did not had all the information found in `PortState`, so I wanted to use with https://github.com/oxidecomputer/hubris/blob/f065736a3502e07f072422af60db141308536e91/drv/ignition-api/src/lib.rs#L602 in an upcoming Hiffy function. But I can get that data from `Port` now, so there indeed is no need for this. I'll rework this.